### PR TITLE
feat: Added priority queue using STL

### PR DESCRIPTION
**Pull request by ttuanho**
#### Description of Change

Added priority queue using STL

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1094"></a>

### CODE REVIEW

**Comment by ttuanho** [comment]
The formatter is so bizzarre. Why does it throw this error? Isn't this std lib?

```
/home/runner/work/C-Plus-Plus/C-Plus-Plus/data_structures/priority_queue.h:12:10: error: 'cassert' file not found [clang-diagnostic-error]
#include <cassert>
         ^
```

**Comment by Panquesito7** [comment]
> Appreciate if someone can explain || help me with this.
> 
> The formatter is so bizzarre to me. Why does it throw this error? Isn't this std lib?
> 
> ```bash
> /home/runner/work/C-Plus-Plus/C-Plus-Plus/data_structures/priority_queue.h:12:10: error: 'cassert' file not found [clang-diagnostic-error]
> #include <cassert>
>          ^
> ```

The file's format shall be `.hpp` if you're using the C++ language.
For the C language, you must use `.h` format.

File: data_structures/test_priority_queue.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+ * @see test_queue.c
+ */
+
+#include <cassert>
+#include <iostream>
+#include <string>
+
+#include "./priority_queue.h"
+
+int main() {
```
Comment (Panquesito7) [inline data_structures/test_priority_queue.cpp:24]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
int main() {
```

File: data_structures/priority_queue.h [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+    /** Clear queue */
+    void clear() { queueFront = NULL; }
+
+    /** Get the size of the queue*/
+    int getSize() { return size; }
+
+ private:
+    node<Kind, Priority> *queueFront; /**< Pointer to the front of the queue */
+    node<Kind, Priority> *queueRear;  /**< Pointer to the rear of the queue  */
```
Comment (Panquesito7) [inline data_structures/priority_queue.h]:
```suggestion
    node<Kind, Priority> *queueFront; ///< Pointer to the front of the queue
    node<Kind, Priority> *queueRear;  ///< Pointer to the rear of the queue
```

File: data_structures/priority_queue.h [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    int size;
```
Comment (Panquesito7) [inline data_structures/priority_queue.h]: Missing variable documentation.

Comment (ttuanho) [reply data_structures/priority_queue.h]: I'm not sure what to add here. This is just the *invariant that keeps track the size of the queue*. Should I add this description as in-line comment?

**Comment by ttuanho** [comment]
> > Appreciate if someone can explain || help me with this.
> > The formatter is so bizzarre to me. Why does it throw this error? Isn't this std lib?
> > ```shell
> > /home/runner/work/C-Plus-Plus/C-Plus-Plus/data_structures/priority_queue.h:12:10: error: 'cassert' file not found [clang-diagnostic-error]
> > #include <cassert>
> >          ^
> > ```
> 
> The file's format shall be `.hpp` if you're using the C++ language.
> For the C language, you must use `.h` fo

> > Appreciate if someone can explain || help me with this.
> > The formatter is so bizzarre to me. Why does it throw this error? Isn't this std lib?
> > ```shell
> > /home/runner/work/C-Plus-Plus/C-Plus-Plus/data_structures/priority_queue.h:12:10: error: 'cassert' file not found [clang-diagnostic-error]
> > #include <cassert>
> >          ^
> > ```
> 
> The file's format shall be `.hpp` if you're using the C++ language.
> For the C language, you must use `.h` format.

Oops. Tyvm. Just updated.

File: data_structures/priority_queue.h (hunk 3 above)
Comment (Panquesito7) [reply data_structures/priority_queue.h]: Sure, sounds good. 👍

File: data_structures/test_priority_queue.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+ * </ul>
+ * @note
+ * @see test_queue.c
+ */
+
+#include <cassert>
+#include <iostream>
+#include <string>
+
+#include "./priority_queue.h"
```
Comment (Panquesito7) [inline data_structures/test_priority_queue.cpp:22]: This shall fix `clang-tidy` warnings.
```suggestion
#include "./priority_queue.hpp"
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: data_structures/priority_queue.hpp [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+/** Definition of the priority queue class */
+template <class Kind, class Priority>
+class queue {
+ public:
+    /** Show queue */
+    void display() {
+        node<Kind, Priority> *current = queueFront;
+        std::cout << "Front --> ";
+        while (current != NULL) {
```
Comment (Panquesito7) [inline data_structures/priority_queue.hpp:35]: C++ algorithms should always use `nullptr`.
Please check other parts of the code for this.
```suggestion
        while (current != nullptr) {
```

File: data_structures/priority_queue.hpp [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+            if (maxElem == queueFront) {
+                queueFront = queueFront->next;
+            } else if (maxElem == queueRear) {
+                maxPrev->next = NULL;
+                queueRear = maxPrev;
+            } else if (maxPrev != NULL) {
+                maxPrev->next = after;
+                if (after == NULL)
+                    queueRear = maxPrev;
```
Comment (Panquesito7) [inline data_structures/priority_queue.hpp:118]:
```suggestion
                if (after == NULL) {
                    queueRear = maxPrev;
                }
```

File: data_structures/priority_queue.hpp (hunk 5 above)
Comment (ttuanho) [reply data_structures/priority_queue.hpp]: Yup. Just pushed.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/runs/1189224084) warnings.

**Comment by ttuanho** [comment]
The formatting fails due to outdated code of `avltree.cpp` which I did only change `NUL` to `nullptr`. The `PriorityQueue` should have passed `clang-tidy` checks

File: data_structures/priority_queue.hpp [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    /** Destructor */
+    ~queue() = default;
+
+    /** Determine whether the queue is empty */
+    bool isEmptyQueue() { return (queueFront == nullptr); }
+
+    /** Add new item to the queue. */
+    void insert(Kind item, Priority lvl) {
+        // std::unique_ptr<node<Kind, Priority>> newNode = new node<Kind, Priority>;
+        auto newNode = make_unique<node<Kind, Priority>>();
```
Comment (ttuanho) [inline data_structures/priority_queue.hpp:113]: Not available in C++11.

File: data_structures/priority_queue.hpp [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+ * @brief Implementation of [Priority
+ * queue](https://en.wikipedia.org/wiki/Priority_queue) an abstract data type
+ * similar to regular queue or stack data structure in which each element
+ * additionally has a "priority" associated with it.
+ * @note improvised from queue.h
+ * @see queue.h
+ */
+
+#ifndef DATA_STRUCTURES_PRIORITY_QUEUE_H_
+#define DATA_STRUCTURES_PRIORITY_QUEUE_H_
```
Comment (Panquesito7) [inline data_structures/priority_queue.hpp:13]: I think these must be `HPP`.
```suggestion
#ifndef DATA_STRUCTURES_PRIORITY_QUEUE_HPP_
#define DATA_STRUCTURES_PRIORITY_QUEUE_HPP_
```

File: data_structures/priority_queue.hpp [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    /** Get the size of the queue*/
+    int getSize() { return size; }
+
+ private:
+    node<Kind, Priority> *queueFront; // < Pointer to the front of the queue 
+    node<Kind, Priority> *queueRear;  // < Pointer to the rear of the queue  
+    int size;                         // < Invariant that keeps track the size of the queue
+};
+
+#endif  // DATA_STRUCTURES_PRIORITY_QUEUE_H_
```
Comment (Panquesito7) [inline data_structures/priority_queue.hpp:195]:
```suggestion
#endif  // DATA_STRUCTURES_PRIORITY_QUEUE_HPP_
```

File: .gitignore [hunk 10]
```diff
@@ -34,4 +34,6 @@ a.out
 *.out
 *.app
 
+bin/
```
Comment (Panquesito7) [inline .gitignore:37]: These changes should be reverted (I think).

**Pull request closed without merging**